### PR TITLE
JAX implementation of scipy.stats.vonmises pdf and logpdf

### PR DESCRIPTION
**Pull request by harryjulian**
Added some of the von Mises functions (as mentioned in #11737) as well as respective tests.

Edit: Fixed tests.

### CODE REVIEW

**Review by froystig — CHANGES_REQUESTED** [review]
Thanks for the PR! I made some minor edit suggestions below. Also, could you [squash together the commits](https://jax.readthedocs.io/en/latest/contributing.html#single-change-commits-and-pull-requests) on this PR branch?

**Comment by harryjulian** [comment]
I'll squash the commits tomorrow!

**Comment by harryjulian** [comment]
I think this should be done, let me know if I need to make any other changes @froystig

**Review by froystig — APPROVED** [review]

File: jax/_src/scipy/stats/vonmises.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+from jax._src.numpy import lax_numpy as jnp
+from jax._src.numpy.util import _wraps, _promote_args_inexact
+from jax._src.typing import Array, ArrayLike
+
+def _cosm1(x: ArrayLike) -> Array:
+  minus_one = _lax_const(x, -1)
+  return lax.add(lax.cos(x), minus_one)
+
+def _i0e(x: ArrayLike) -> Array:
+  return lax.exp(-lax.abs(x)) * jnp.i0(x)
```
Comment (jakevdp) [inline jax/_src/scipy/stats/vonmises.py:28]: I think this utility could be replaced with a call to `lax.bessel_i0e`

File: jax/_src/scipy/stats/vonmises.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+def _pdf(x: ArrayLike, kappa: ArrayLike) -> Array:
+  pi = _lax_const(kappa, jnp.pi)
+  return lax.exp(kappa * _cosm1(x)) / (2 * pi * _i0e(kappa))
```
Comment (jakevdp) [inline jax/_src/scipy/stats/vonmises.py:32]: Hi - it looks like there are some errors when `JAX_ENABLE_X64=1` (you should be able to see the "Tests with internal numpy/scipy" failures in the CI list) One of these binary operations is happening between two values that have mismatched types. We check for this in the test suite in order to support `jax_numpy_dtype_promotion='strict'` without false positives. I think the issue is that `lax_const` returns a numpy scalar, and so `2 * pi` is a numpy operation that promotes to float64. If you avoid the const and just write `2 * jnp.pi * lax.bessel_i0e(kappa)` I think it should fix the issue.

**Review by jakevdp — CHANGES_REQUESTED** [review]

**Comment by harryjulian** [comment]
Updated!

**Review by jakevdp — CHANGES_REQUESTED** [review]
One more thing, sorry I didn't notice it earlier.

Also, please add these new functions to the docs at https://github.com/google/jax/blob/main/docs/jax.scipy.rst#jaxscipystats; you can follow the pattern of the other distributions listed there.

File: jax/_src/scipy/stats/vonmises.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+@_wraps(osp_stats.vonmises.pdf, update_doc=False)
+def pdf(x: ArrayLike, kappa: ArrayLike) -> Array:
+  x, kappa = _promote_args_inexact('vonmises.pdf', x, kappa)
+  zero = _lax_const(kappa, 0)
+  return jnp.where(lax.gt(kappa, zero), _pdf(x, kappa), jnp.nan)
+
+@_wraps(osp_stats.vonmises.logpdf, update_doc=False)
+def logpdf(x: ArrayLike, kappa: ArrayLike) -> Array:
+  return lax.log(pdf(x, kappa))
```
[the remark this answers is not in the document]
Comment (harryjulian) [reply jax/_src/scipy/stats/vonmises.py:34]: I've made the other updates, but I'm not sure what you mean here. Are you referring to the first term of the pdf calculation? I'm not sure where the exponent of the logpdf is being taken.

[the remark this answers is not in the document]
Comment (jakevdp) [reply jax/_src/scipy/stats/vonmises.py:34]: In the `_pdf` function, you compute `kappa * (jnp.cos(x) - 1) ) / (2 * jnp.pi * lax.bessel_i0e(kappa)`, which is essentially `logpdf`. You then pass it to `lax.exp` in order to compute the pdf, then here pass that result to `lax.log`, so the return value here is essentially `log(exp(logpdf))`. Because floating point arithmetic is not exact, it would be better to avoid taking the log of the exponent of the result that you want to return. Does that make sense?

[the remark this answers is not in the document]
Comment (jakevdp) [reply jax/_src/scipy/stats/vonmises.py:34]: You can see this pattern used for other distributions, for example here: https://github.com/google/jax/blob/08b6b0dd43674da97ca00b957d16488845e7931f/jax/_src/scipy/stats/norm.py#L28-L39

[the remark this answers is not in the document]
Comment (harryjulian) [reply jax/_src/scipy/stats/vonmises.py:34]: `lax.exp` doesn't completely wrap the entirety of `kappa * (jnp.cos(x) - 1) ) / (2 * jnp.pi * lax.bessel_i0e(kappa)` - only `kappa * (jnp.cos(x) - 1)` is passed to it. It's part of the actual pdf calculation on all of the examples I can see on the internet, so I don't think it's computing the exponent of the logpdf

i.e. [link](https://mathworld.wolfram.com/vonMisesDistribution.html)

[the remark this answers is not in the document]
Comment (jakevdp) [reply jax/_src/scipy/stats/vonmises.py:34]: I see, sorry I misread the computation.

Still, I think we need to change this to compute logpdf directly, and then implement `pdf` by computing `exp(logpdf)`. The reason `logpdf` exists at all is because computing `logpdf` using `log(pdf)` in general loses accuracy for small probabilities, and accurate logs of small probabilities are important for many applications.

To be more explicit, I think the code should look something like this (untested):
```python
@_wraps(osp_stats.vonmises.pdf, update_doc=False)
def pdf(x: ArrayLike, kappa: ArrayLike) -> Array:
  return jnp.exp(logpdf(x, kappa))

@_wraps(osp_stats.vonmises.logpdf, update_doc=False)
def logpdf(x: ArrayLike, kappa: ArrayLike) -> Array:
  x, kappa = _promote_args_inexact('vonmises.logpdf', x, kappa)
  return jnp.where(kappa > 0, kappa * (jnp.cos(x) - 1) - jnp.log(2 * jnp.pi * lax.bessel_i0e(kappa)), jnp.nan)
```

[the remark this answers is not in the document]
Comment (harryjulian) [reply jax/_src/scipy/stats/vonmises.py]: This makes sense. Altered, but not sure why the docs aren't building.

File: docs/jax.scipy.rst [hunk 4]
```diff
@@ -355,3 +355,12 @@ jax.scipy.stats.gaussian_kde
    gaussian_kde.resample
    gaussian_kde.pdf
    gaussian_kde.logpdf
+
+jax.scipy.stats.vonmises
+~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (jakevdp) [inline docs/jax.scipy.rst:360]: The length of this underline needs to match the length of the text above it

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks great!

To make the docs build pass, please rebase on an updated main branch. There was an ipython version dependency issue that should already be fixed there.

**Comment by harryjulian** [comment]
Sorted! Thanks for the guidance.

**Review by jakevdp — APPROVED** [review]
Looks great! Thanks for sticking with this through all the changes.

Docs seem to be building without an issue: https://jax--13355.org.readthedocs.build/en/13355/jax.scipy.html?highlight=vonmises#module-jax.scipy.stats.vonmises
